zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -834,30 +834,31 @@ public int aclCacheSize() {
 
         public List<ProcessTxnResult> multiResult;
 
-        /**
-         * Equality is defined as the clientId and the cxid being the same. This
-         * allows us to use hash tables to track completion of transactions.
-         *
-         * @see java.lang.Object#equals(java.lang.Object)
-         */
         @Override
-        public boolean equals(Object o) {
-            if (o instanceof ProcessTxnResult) {
-                ProcessTxnResult other = (ProcessTxnResult) o;
-                return other.clientId == clientId && other.cxid == cxid;
-            }
+        public boolean equals(Object obj) {
+          if (this == obj) {
+            return true;
+          }
+          if (!(obj instanceof ProcessTxnResult)) {
             return false;
+          }
+          ProcessTxnResult other = (ProcessTxnResult) obj;
+          if (clientId != other.clientId) {
+            return false;
+          }
+          if (cxid != other.cxid) {
+            return false;
+          }
+          return true;
         }
 
-        /**
-         * See equals() to find the rational for how this hashcode is generated.
-         *
-         * @see ProcessTxnResult#equals(Object)
-         * @see java.lang.Object#hashCode()
-         */
         @Override
         public int hashCode() {
-            return (int) ((clientId ^ cxid) % Integer.MAX_VALUE);
+          final int prime = 31;
+          int result = 1;
+          result = prime * result + (int) (clientId ^ (clientId >>> 32));
+          result = prime * result + cxid;
+          return result;
```

Comment:
Why not something simple, like `Long.hashCode(clientId + cxid)`?

Reply:
This proposed solution is not great for collisions, since the two values will collide (For example):

clientId = 100
cxid = 0

clientId = 99
cxid = 1

Better to treat them individually. I did simplify with Long#hashCode